src/components/grouped-character/grouped-character.test.tsx

```diff
@@ -0,0 +1,451 @@
+      groups={[2, 2, 4]}
+      onChange={onChange}
+    />
+  );
+
+  await user.type(screen.getByRole("textbox"), "cc-aa-aabb");
+
+  expect(onChange).toHaveBeenLastCalledWith(
+    expect.objectContaining({
+      target: expect.objectContaining({
+        value: {
+          formattedValue: "cc-aa-aabb",
+          rawValue: "ccaaaabb",
+        },
+      }),
+    })
+  );
+});
+
+test("the component takes configuration for how characters should be grouped", () => {
+  render(
+    <GroupedCharacter separator="-" value="12345678" groups={[2, 2, 4]} />
+  );
+
+  expect(screen.getByRole("textbox")).toHaveValue("12-34-5678");
+});
+
+test("emits a formatted string on change event", async () => {
+  const onChange = jest.fn();
+  const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+  render(<MockComponent initialValue="" onChange={onChange} />);
+
+  const input = screen.getByRole("textbox");
+
+  await user.click(input);
+  await user.clear(input);
+  await user.type(input, "123456");
+
+  expect(onChange).toHaveBeenLastCalledWith(
+    expect.objectContaining({
+      target: expect.objectContaining({
+        value: expect.objectContaining({
+          formattedValue: "12-34-56",
+          rawValue: "123456",
+        }),
+      }),
+    })
+  );
+});
+
+test("invokes provided onChange handler with proper event target name and id if those are provided", async () => {
+  const onChange = jest.fn();
+  const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+  render(
+    <MockComponent
+      initialValue="12345678"
+      onChange={(ev) => {
+        onChange?.({
+          value: ev.target.value,
+          id: "unique_id",
+          name: "nice_name",
+        });
+      }}
+    />
+  );
+
+  const input = screen.getByRole("textbox");
+
+  await user.click(input);
+  await user.clear(input);
+  await user.type(input, "1");
+
+  expect(onChange).toHaveBeenLastCalledWith(
+    expect.objectContaining({
+      id: "unique_id",
+      name: "nice_name",
+      value: {
+        formattedValue: "1",
+        rawValue: "1",
+      },
+    })
+  );
+});
+
+test("emits a formatted string on blur event", async () => {
+  const onBlur = jest.fn();
+  const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });
+
+  render(
+    <MockComponent
+      initialValue="12345678"
+      onBlur={(ev) => {
+        onBlur?.(ev.target.value);
+      }}
+    />
+  );
+
+  const input = screen.getByRole("textbox");
+
+  await user.click(input);
+  await user.clear(input);
+  await user.type(screen.getByRole("textbox"), "123456");
+  await user.tab();
+
+  expect(onBlur).toHaveBeenLastCalledWith(
+    expect.objectContaining({
+      formattedValue: "12-34-56",
+      rawValue: "123456",
+    })
+  );
+});
+
+// coverage
+// This test scenario may seem obvious but it is required for coverage
+test("does nothing if onBlur is not provided", async () => {
+  const onChange = jest.fn();
```

Comment:
**suggestion** (non-blocking): I wasn't going to bother to leave this comment, since the test is a useless one purely for coverage, but since I've ended up leaving another comment I thought I might as well...

this is another `onChange` that doesn't really have anything to do with the behaviour being tested. The "relevant" thing here is the `onBlur` (and its lack of being passed to the component), whether `onChange` is passed or not seems irrelevant.

Unless of course that turns out to be needed for coverage, in which case no bother - but I'd be very surprised.